lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// resolveReqResult categorizes a status code between serverError, clientError and success.
+// The status code could be coming from an asynchronous operation or a request response.
+func resolveReqResult(statusCode int, asyncOperation bool) requestResult {
+	// Status codes from operations follow a different pattern.
+	if asyncOperation {
+		// If status is not 200(succeeded) nor 401(cancelled), it is considered a failure.
+		if statusCode != 200 && statusCode != 401 {
+			return errorServer
+		}
+
+		return success
+	}
+
+	// 4* codes are considered client errors on HTTP.
+	if statusCode >= 400 && statusCode < 500 {
+		return errorClient
+	}
+
+	// 2* codes are considered successes on HTTP, 300 are for forwarded requests
+	// that are included as successes for the purpose of the metrics.
+	if statusCode >= 200 && statusCode < 400 {
+		return success
+	}
+
+	return errorServer
+}
+
+// resolveEntityType derives the main entity type an endpoint is opertaing on from its URL.
+// This is used to categorize completed requests in internal metrics.
+func resolveEntityType(url url.URL) entity.Type {
+	path := url.Path
+	if url.RawPath != "" {
+		path = url.RawPath
+	}
+
+	// For /, /{version} and /internal endpoints.
+	if !strings.HasPrefix(path, "/"+version.APIVersion+"/") {
+		return entity.TypeServer
+	}
+
+	pathParts := strings.Split(strings.TrimPrefix(path, "/"+version.APIVersion+"/"), "/")
+
+	switch pathParts[0] {
+	// For /{version}/containers, /{version}/virtual-machines and /{version}/instances endpoints.
+	case "containers", "virtual-machines", "instances":
+		return entity.TypeInstance
+	// For /{version}/network-zones, /{version}/network-allocations, /{version}/network-acls and /{version}/networks endpoints.
+	case "network", "network-acls", "network-allocations", "network-zones":
+		return entity.TypeNetwork
+	// For /{version}/storage-pools and /{version}/storage-volumes endpoints.
+	case "storage-pools", "storage-volumes":
+		return entity.TypeStoragePool
+	// For /{version}/auth and /{version}/certificates endpoints.
+	case "auth", "certificates":
+		return entity.TypeIdentity
+	// For /{version}/images endpoints.
+	case "images":
+		return entity.TypeImage
+	// For /{version}/cluster endpoints.
+	case "cluster":
+		return entity.TypeNode
+	// For /{version}/projects endpoints.
+	case "projects":
+		return entity.TypeProject
+	// For /{version}/profiles endpoints.
+	case "profiles":
+		return entity.TypeProfile
+	// For /{version}/warnings endpoints.
+	case "warnings":
+		return entity.TypeWarning
+	// For /{version}/operations endpoints.
+	case "operations":
+		return entity.TypeOperation
+	// For /{version}/events, /{version}/metrics, /{version}/metadata/configuration and /{version}/resources endpoints.
+	default:
+		return entity.TypeServer
+	}
+}
+
+var apiMetricsEntityTypes = []entity.Type{
+	entity.TypeInstance,
+	entity.TypeNetwork,
+	entity.TypeStoragePool,
+	entity.TypeIdentity,
+	entity.TypeImage,
+	entity.TypeNode,
```

Comment:
@markylaing @hamistao where is this entity type used? We shouldnt be exposing "node" as an entity type as the user facing concept is "cluster member". Can we change this or does it break the ACL?

Reply:
@hamistao I feel uncomfortable about having this list, and the prefixes above in `resolveEntityType` defined in the `metrics` package as I suspect this will easily fall out of date when we add new entity types.

@markylaing can we change `shared/entity/type.entityTypes` to a map keyed on each `Type` and value that is an unexported struct that contains a field with URL prefixes that are used to match to that entity type.

Then we can add a function like `resolveEntityType` called `ParseURLPrefix` that just returns the EntityType or empty string for unrecognised ones.

Then this `metrics` package can pass the endpoint URL to it, get an EntityType back, and then in the `metrics` package we just have the metric specific grouping logic (i,e which URL prefixes we are grouping together that do have their own entity types).

Reply:
@tomponline I suspect this is why we need to have the entity type refactor sorted sooner rather than later. Any time we need to add a general concept to LXD we're going to run into this issue.

I like the idea of having a map that returns a struct, we can define this struct for each type and have it implement an interface so it's clear what needs to be added if adding a new entity type.

Reply:
yeah agreed, shall we catch up about that in our 1:1 next week?

Reply:
I think the main concern was around the interface embedding interfaces.

Reply:
I'd like to have a quick play with it to see if I can make it more intuitive but yes let's catch up about it in our 1:1. If something more important crops up lets plan a dedicated meeting.

Reply:
@tomponline 
For the purposes of the API rates metrics, should I leave the entity type resolution as is until you have sorted this refactor?
If so, I will also wait for this before implementing the requested unit tests for this function.

Not sure about this, but instead of defining a concept of entity type groups on the `metrics` package, would it make sense to include this concept on the refactor?

As for `entity.TypeNode`, I could use `entity.TypeClusterGroup`. we discussed this today and it could be changed without any side effects, I just don't think it is appropriate to include the change on this PR.

@markylaing If needed, I can contribute to the refactor myself with the bits needed for the API metrics.